### DISCUSSION TITLE
Difficulty in Writing unit test .

### DISCUSSION THREAD

**Discussion by hoBabu1** [Cyfrin/foundry-full-course-cu#1970] [Q&A] [answered]
![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/128292819/adf26c98-29a4-4649-9da6-93f19cadba49)
Could you please guide me how can i write test for ```  bool success = IERC20(tokenColletral).transferFrom(msg.sender, address(this), amount);
        if (!success) {
            revert DSCBrain__TransferOfTokenFailedFromUserToContract();
        }``` .

**Comment by EngrPips** [comment]
Hello @hoBabu1, Doing this would require a whole different setup, meaning you would need to use an `ERC20` token that would revert on `transferFrom` call. You can look into the GitHub `https://github.com/Cyfrin/foundry-defi-stablecoin-cu/blob/main/test/unit/DSCEngineTest.t.sol` associated with the DEFI action to see how Patrick implemented it

**Reply by hoBabu1** [reply]
> Hello @hoBabu1, Doing this would require a whole different setup, meaning you would need to use an `ERC20` token that would revert on `transferFrom` call. You can look into the GitHub `https://github.com/Cyfrin/foundry-defi-stablecoin-cu/blob/main/test/unit/DSCEngineTest.t.sol` associated with the DEFI action to see how Patrick implemented it

But i am not able to understand that . 
could you please explain that once .

**Reply by EngrPips** [reply, marked as answer]
Alright, I will paste the function for the test here and write comments on each line. Let's see if that makes you understand.

```js

// this test needs it's own setup
    function testRevertsIfTransferFromFails() public {
        // Arrange - Setup
        address owner = msg.sender;
        vm.prank(owner);
        // set up a new token that will revert when we call `transferFrom` on it
        MockFailedTransferFrom mockDsc = new MockFailedTransferFrom();
       // we pass in the token to the `tokenAddresses` so that we can use it to test `bool success = 
       // IERC20(tokenColletral).transferFrom(msg.sender, address(this), amount); if (!success) { revert 
       // DSCBrain__TransferOfTokenFailedFromUserToContract(); }`
        tokenAddresses = [address(mockDsc)];
       // we set a priceFeed for the token
        feedAddresses = [ethUsdPriceFeed];
        vm.prank(owner);
       // we deploy a new `DSCEngine` contract passing in the token which will revert when we call `transferFrom` on it
        DSCEngine mockDsce = new DSCEngine(tokenAddresses, feedAddresses, address(mockDsc));
       // we mint some amount of the token to `user` address
        mockDsc.mint(user, amountCollateral);

        vm.prank(owner);
       // we transfer the ownership of the token to the `DSCEngine` so that the `DSCEngine` can have control over the token
        mockDsc.transferOwnership(address(mockDsce));
        // Arrange - User
        vm.startPrank(user);
       // The user approve the `DSCEngine` to take their token
        ERC20Mock(address(mockDsc)).approve(address(mockDsce), amountCollateral);
        // Act / Assert
       // we tell the test to expect a revert
        vm.expectRevert(DSCEngine.DSCEngine__TransferFailed.selector);
       // The user called `depositCollateral` function on the `DSCEngine` which will call the `transferFrom` function on the `token` 
      // we have created to revert whenever `transferFrom` is called on it. So the process would revert and our test will pass
        mockDsce.depositCollateral(address(mockDsc), amountCollateral);
        vm.stopPrank();
    }
```
Please take your time to read the comment in the function above and ask questions if anything is confusing.

**Reply by hoBabu1** [reply]
> Alright, I will paste the function for the test here and write comments on each line. Let's see if that makes you understand.
> 
> ```js
> // this test needs it's own setup
>     function testRevertsIfTransferFromFails() public {
>         // Arrange - Setup
>         address owner = msg.sender;
>         vm.prank(owner);
>         // set up a new token that will revert when we call `transferFrom` on it
>         MockFailedTransferFrom mockDsc = new MockFailedTransferFrom();
>        // we pass in the token to the `tokenAddresses` so that we can use it to test `bool success = 
>        // IERC20(tokenColletral).transferFrom(msg.sender, address(this), amount); if (!success) { revert 
>        // DSCBrain__TransferOfTokenFailedFromUserToContract(); }`
>         tokenAddresses = [address(mockDsc)];
>        // we set a priceFeed for the token
>         feedAddresses = [ethUsdPriceFeed];
>         vm.prank(owner);
>        // we deploy a new `DSCEngine` contract passing in the token which will revert when we call `transferFrom` on it
>         DSCEngine mockDsce = new DSCEngine(tokenAddresses, feedAddresses, address(mockDsc));
>        // we mint some amount of the token to `user` address
>         mockDsc.mint(user, amountCollateral);
> 
>         vm.prank(owner);
>        // we transfer the ownership of the token to the `DSCEngine` so that the `DSCEngine` can have control over the token
>         mockDsc.transferOwnership(address(mockDsce));
>         // Arrange - User
>         vm.startPrank(user);
>        // The user approve the `DSCEngine` to take their token
>         ERC20Mock(address(mockDsc)).approve(address(mockDsce), amountCollateral);
>         // Act / Assert
>        // we tell the test to expect a revert
>         vm.expectRevert(DSCEngine.DSCEngine__TransferFailed.selector);
>        // The user called `depositCollateral` function on the `DSCEngine` which will call the `transferFrom` function on the `token` 
>       // we have created to revert whenever `transferFrom` is called on it. So the process would revert and our test will pass
>         mockDsce.depositCollateral(address(mockDsc), amountCollateral);
>         vm.stopPrank();
>     }
> ```
> 
> Please take your time to read the comment in the function above and ask questions if anything is confusing.

Got it . 
Thanks a lot sir

**Reply by EngrPips** [reply]
I am very Glad you got it. Keep crushing it, Friend.
